File: internal/lambda/strategies/delay.go
package strategies

import (
	"context"
	"crypto/rand"
	"fmt"
	"math/big"
	"sync"
	"time"

	"github.com/coachpo/meltica/internal/schema"
)

// Delay simulates processing latency without performing any business logic.
type Delay struct {
	MinDelay time.Duration
	MaxDelay time.Duration
}

const (
	// DefaultMinDelay defines the default lower bound for random delay.
	DefaultMinDelay = 100 * time.Millisecond
	// DefaultMaxDelay defines the default upper bound for random delay.
	DefaultMaxDelay = 500 * time.Millisecond
)

var (
	delaySubscribedEvents = []schema.RouteType{
		schema.RouteTypeTrade,
		schema.RouteTypeTicker,
		schema.RouteTypeOrderbookSnapshot,
		schema.RouteTypeAccountBalance,
	}
	delayRandMu sync.Mutex
)

// SubscribedEvents returns the list of event types this strategy subscribes to.
func (s *Delay) SubscribedEvents() []schema.RouteType {
	return append([]schema.RouteType(nil), delaySubscribedEvents...)
}

func (s *Delay) sleep() {
	minDelay := s.MinDelay
	maxDelay := s.MaxDelay
	if minDelay == 0 && maxDelay == 0 {
		minDelay = DefaultMinDelay
		maxDelay = DefaultMaxDelay
	}
	if minDelay < 0 {
		minDelay = 0
	}
	if maxDelay < 0 {
		maxDelay = 0
	}
	if maxDelay < minDelay {
		maxDelay = minDelay
	}

	delayRandMu.Lock()
	defer delayRandMu.Unlock()

	delayRange := maxDelay - minDelay
	if delayRange <= 0 {
		time.Sleep(minDelay)
		return
	}

	offset, err := randDuration(delayRange)
	if err != nil {
		// Fallback to midpoint delay if crypto/rand fails
		time.Sleep(minDelay + delayRange/2)
		return
	}
	time.Sleep(minDelay + offset)
}

func randDuration(maxDuration time.Duration) (time.Duration, error) {
	if maxDuration <= 0 {
		return 0, nil
	}
	nanos := maxDuration.Nanoseconds()
	if nanos <= 0 {
		return 0, nil
	}
	limit := big.NewInt(nanos + 1)
	value, err := rand.Int(rand.Reader, limit)
	if err != nil {
		return 0, fmt.Errorf("crypto rand int: %w", err)
	}
	return time.Duration(value.Int64()), nil
}

// OnTrade handles trade events by adding a delay.
func (s *Delay) OnTrade(_ context.Context, _ *schema.Event, _ schema.TradePayload, _ float64) {
	s.sleep()
}

// OnTicker handles ticker events by adding a delay.
func (s *Delay) OnTicker(_ context.Context, _ *schema.Event, _ schema.TickerPayload) {
	s.sleep()
}

// OnBookSnapshot handles order book snapshot events by adding a delay.
func (s *Delay) OnBookSnapshot(_ context.Context, _ *schema.Event, _ schema.BookSnapshotPayload) {
	s.sleep()
}

// OnOrderFilled handles order fill events by adding a delay.
func (s *Delay) OnOrderFilled(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
	s.sleep()
}

// OnOrderRejected handles order rejection events by adding a delay.
func (s *Delay) OnOrderRejected(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload, _ string) {
	s.sleep()
}

// OnOrderPartialFill handles order partial fill events by adding a delay.
func (s *Delay) OnOrderPartialFill(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
	s.sleep()
}

// OnOrderCancelled handles order cancellation events by adding a delay.
func (s *Delay) OnOrderCancelled(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
	s.sleep()
}

// OnOrderAcknowledged handles order acknowledgment events by adding a delay.
func (s *Delay) OnOrderAcknowledged(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
	s.sleep()
}

// OnOrderExpired handles order expiration events by adding a delay.
func (s *Delay) OnOrderExpired(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
	s.sleep()
}

// OnKlineSummary handles kline summary events by adding a delay.
func (s *Delay) OnKlineSummary(_ context.Context, _ *schema.Event, _ schema.KlineSummaryPayload) {
	s.sleep()
}

// OnInstrumentUpdate handles instrument catalogue refreshes by adding a delay.
func (s *Delay) OnInstrumentUpdate(_ context.Context, _ *schema.Event, _ schema.InstrumentUpdatePayload) {
	s.sleep()
}

// OnBalanceUpdate handles balance updates by adding a delay.
func (s *Delay) OnBalanceUpdate(_ context.Context, _ *schema.Event, _ schema.BalanceUpdatePayload) {
	s.sleep()
}

File: strategies/delay/v1.0.0/delay.js
const DEFAULT_MIN_DELAY = "100ms";
const DEFAULT_MAX_DELAY = "500ms";

function parseDuration(value, fallback) {
  if (value === undefined || value === null || value === "") {
    return fallback;
  }
  const text = String(value).trim().toLowerCase();
  if (text === "") {
    return fallback;
  }
  const match = text.match(/^(-?\d+(?:\.\d+)?)(ns|us|µs|ms|s|m|h)?$/);
  if (!match) {
    return fallback;
  }
  const magnitude = Number(match[1]);
  if (!Number.isFinite(magnitude) || magnitude < 0) {
    return fallback;
  }
  const unit = match[2] || "ms";
  switch (unit) {
    case "ns":
      return magnitude / 1_000_000;
    case "us":
    case "µs":
      return magnitude / 1_000;
    case "ms":
      return magnitude;
    case "s":
      return magnitude * 1_000;
    case "m":
      return magnitude * 60_000;
    case "h":
      return magnitude * 3_600_000;
    default:
      return fallback;
  }
}

function clampDurations(minMs, maxMs) {
  let min = Math.max(0, minMs);
  let max = Math.max(0, maxMs);
  if (max < min) {
    max = min;
  }
  return { min, max };
}

function randomDelay(minMs, maxMs) {
  if (maxMs <= minMs) {
    return minMs;
  }
  const offset = Math.random() * (maxMs - minMs);
  return minMs + offset;
}

module.exports = {
  metadata: {
    name: "delay",
    version: "1.0.0",
    displayName: "Delay",
    description: "Simulates processing latency with a configurable random delay window.",
    config: [
      {
        name: "min_delay",
        type: "duration",
        description: "Lower bound for the random delay interval",
        default: DEFAULT_MIN_DELAY,
        required: false
      },
      {
        name: "max_delay",
        type: "duration",
        description: "Upper bound for the random delay interval",
        default: DEFAULT_MAX_DELAY,
        required: false
      },
      {
        name: "dry_run",
        type: "bool",
        description: "When true, strategy logs intended orders without submitting them",
        default: true,
        required: false
      }
    ],
    events: [
      "Trade",
      "Ticker",
      "BookSnapshot",
      "KlineSummary",
      "BalanceUpdate",
      "RiskControl",
      "InstrumentUpdate",
      "ExecReport"
    ]
  },
  create: function (env) {
    const minConfig = parseDuration(env.config.min_delay ?? DEFAULT_MIN_DELAY, 100);
    const maxConfig = parseDuration(env.config.max_delay ?? DEFAULT_MAX_DELAY, 500);
    const { min, max } = clampDurations(minConfig, maxConfig);

    function sleepRandom() {
      const durationMs = randomDelay(min, max);
      env.helpers.sleep(durationMs);
    }

    return {
      wantsCrossProviderEvents: function () {
        return true;
      },
      onTrade: function () {
        sleepRandom();
      },
      onTicker: function () {
        sleepRandom();
      },
      onBookSnapshot: function () {
        sleepRandom();
      },
      onOrderFilled: function () {
        sleepRandom();
      },
      onOrderRejected: function () {
        sleepRandom();
      },
      onOrderPartialFill: function () {
        sleepRandom();
      },
      onOrderCancelled: function () {
        sleepRandom();
      },
      onOrderAcknowledged: function () {
        sleepRandom();
      },
      onOrderExpired: function () {
        sleepRandom();
      },
      onKlineSummary: function () {
        sleepRandom();
      },
      onInstrumentUpdate: function () {
        sleepRandom();
      },
      onBalanceUpdate: function () {
        sleepRandom();
      },
      onRiskControl: function () {
        sleepRandom();
      }
    };
  }
};
